<a href="https://colab.research.google.com/github/gcasar1/fiap_/blob/main/deteccao_liveness_notebook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# MBA FIAP Inteligência Artificial & Machine Learning

## Visão Computacional: Análise de Imagens Médicas

> Atenção: este notebook foi desenhado para funcionar no **Google Collab**.


## 1. Introdução

Uma determinada fintech focada em consumidores finais pessoa física constataou um grande número de fraudes em transações bancárias.

O setor de fraudes apontou que existem clientes que se queixaram de não contratar serviços específicos, como o crédito pessoal, e após isso transferir para outras contas desconhecidas.

Após análises pelas equipes de segurança, os protocolos de utilização da senha foram realizados em conformidade, ou seja, cada cliente autenticou com sua própria senha de maneira regular.

Em função disso, o banco precisa arcar com reembolsos e medidas de contenção para evitar processos judiciais, pois os clientes alegam terem sido invadidos por hackers ou algo parecido.

Uma das formas de solucionar ou minimizar este problema é com a utilização de outras formas de autenticação, sobretudo em operações críticas, como a obtenção de crédito pessoal.

Desta forma podemos implementar uma verificação de identidade com prova de vida (liveness), que utilize uma verificação e identificação facial.

Caso o cliente não seja autenticado, ele será atendido por uma esteira dedicada e as evidências da não identificação serão encaminhadas para a área de IA para validação dos parâmetros e limiares para aperfeiçoamento do modelo.

Será necessário construir:

* Detector de faces
* Identificação de faces (podendo ser um comparador entre um rosto de documento e outra da prova de vida)
* Detecção de vivacidade (liveness) para evitar que um fraudador utilize uma foto estática.


>Formas alternativas de prover a identificação e prova de vivacidade, além destas que foram solicitadas poderão ser submetidas.


<p align="center">
    <img src="https://github.com/michelpf/fiap-ml-visao-computacional-detector-liveness/blob/master/notebook/imagens/liveness.jpg?raw=1">
</p>

Imagem retirada do [Grunge](https://www.grunge.com/192826/company-testing-robocop-facial-recognition-software-with-us-police/).

## 2. Instruções

Este projeto final tem como objetivo explorar os conhecimentos adquiridos nas aulas práticas.

Iremos constuir uma forma de validar se uma determinada imagem foi ou não adulterada e se trata de uma produção fraudade.

Existem diversas formas de validar a vivacidade, e neste sentido conto com a criatividade de vocês dado que já dominam encontrar uma face numa imagem, aplicar marcos faciais e até mesmo construir uma rede neural convulacional.

A abordagem mais simples é pela construção de uma rede neural com imagens de fotos de rostos de outras fotos e fotos de rostos sem modificações. Tal classificador deverá classificar se dada imagem possui vivacidade ou não com uma pontuação de probabilidade.

Referências que abordam o tema para servir de inspiração:

1. [PyImageSearch](https://pyimagesearch.com/2019/03/11/liveness-detection-with-opencv/), Liveness detection with OpenCV;
2. [Kickertech](https://kickertech.com/face-liveness-detection-via-opencv-and-tensorflow/), Liveness detection via OpenCV and Tensorflow.
3. [Towards Data Science](https://towardsdatascience.com/real-time-face-liveness-detection-with-python-keras-and-opencv-c35dc70dafd3?gi=24f8e1b740f9), Real-time face liveness detection with Python, Keras and OpenCV.

Este projeto poderá ser feita por grupos de até 4 pessoas.
Caso este projeto seja substitutivo, deverá ser realizado por apenas uma pessoa.

| Nome dos Integrantes                       | RM         | Turma   |
| :----------------------------------------- | :--------- | :-----: |
| Gabriela Casari                            | RM 351352  | 5DTSR   |
| Jeferson Souto                             | RM 352350  | 5DTSR   |
| Pedro Henrique Decezaro Vicensi            | RM 351482  | 5DTSR   |

## 3. Abordagem e organização da solução do problema (2 pontos)

Como o grupo pretende deteccar a prova de vivacidade de uma determinada imagem? Quais os passos e os building blocks deste processo?

**Resposta**:
Para detectar a prova de vivacidade em fotos, incluindo tentativas de fraude como o uso de máscaras, o grupo usará o Amazon Rekognition Custom Labels. O modelo será treinado para identificar características sutis que distinguem uma face real de uma falsa, como diferenças de textura e profundidade. O processo envolve a análise de fotos enviadas pelos usuários para detectar anomalias típicas de máscaras ou fotos planas. O Amazon Rekognition facilita a criação de modelos personalizados que podem identificar essas nuances, proporcionando uma camada adicional de segurança em sistemas de autenticação facial. Essa abordagem permite uma detecção rápida e precisa, garantindo que apenas faces reais sejam reconhecidas como válidas.

## 4 Desenvolvimento da solução (5,5 pontos)

Detalhe o passo-a-passo do algoritmo de deteção de vivacidade.
Se optar pela construção e treinamento de um modelo de redes neurais convulucionais, apresente a arquitetura, prepare os dados de treinamento, realize o treinamento.

### 4.1 Organização de dados para treinamento de modelo de liveness (2 pontos)

In [2]:
!pip install boto3

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 139.2/139.2 kB 5.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.4/12.4 MB 60.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 82.7/82.7 kB 4.9 MB/s eta 0:00:00


In [3]:
!pip install awscli

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.5/4.5 MB 32.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 548.2/548.2 kB 29.9 MB/s eta 0:00:00
  Attempting uninstall: rsa
    Found existing installation: rsa 4.9
    Uninstalling rsa-4.9:
      Successfully uninstalled rsa-4.9
  Attempting uninstall: docutils
    Found existing installation: docutils 0.18.1
    Uninstalling docutils-0.18.1:
      Successfully uninstalled docutils-0.18.1


In [4]:
import boto3
import matplotlib.pyplot as plt
import matplotlib.image as mpimg
from PIL import Image, ImageDraw, ImageFont
import json
import requests
import os

In [16]:
ACCESS_KEY = 'xxx'
ACCESS_SECRET = 'xxx'
REGION = "us-east-1"


In [13]:
def start_model(project_arn, model_arn, version_name, min_inference_units):

    client=boto3.client('rekognition')

    try:
        print('Starting model: ' + model_arn)
        response=client.start_project_version(ProjectVersionArn=model_arn, MinInferenceUnits=min_inference_units)
        project_version_running_waiter = client.get_waiter('project_version_running')
        project_version_running_waiter.wait(ProjectArn=project_arn, VersionNames=[version_name])
        describe_response=client.describe_project_versions(ProjectArn=project_arn,
            VersionNames=[version_name])
        for model in describe_response['ProjectVersionDescriptions']:
            print("Status: " + model['Status'])
            print("Message: " + model['StatusMessage'])
    except Exception as e:
        print(e)

    print('Done...')

def main():
    project_arn='arn:aws:rekognition:us-east-1:011528301996:project/Faces/1722552711471'
    model_arn='arn:aws:rekognition:us-east-1:011528301996:project/Faces/version/Faces.2024-08-01T19.59.45/1722553185343'
    min_inference_units=1
    version_name='Faces.2024-08-01T19.59.45'
    start_model(project_arn, model_arn, version_name, min_inference_units)

if __name__ == "__main__":
    main()

Starting model: arn:aws:rekognition:us-east-1:011528301996:project/Faces/version/Faces.2024-08-01T19.59.45/1722553185343
Status: RUNNING
Message: The model is running.
Done...


### 4.2 Treinamento de modelo de liveness (1,5 pontos)

O treinamento foi feito pela Amazon Rekognition Custom.

### 4.3 Métricas de desempenho do modelo (2 pontos)

In [20]:
import boto3
import json

def get_s3_object(bucket, key):
    s3 = boto3.client('s3')
    try:
        response = s3.get_object(Bucket=bucket, Key=key)
        content = response['Body'].read().decode('utf-8')
        return json.loads(content)
    except Exception as e:
        print("Error fetching S3 object:", e)
        return None

def main():
    bucket = 'custom-labels-console-us-east-1-xxx'
    summary_key = 'evaluation/arn:aws:rekognition:us-east-1:011528301996:project/Faces/1722552711471/EvaluationResultSummary-Faces-Faces.2024-08-01T19.59.45.json'

    summary = get_s3_object(bucket, summary_key)

    if summary:
        print("Evaluation Summary:")
        print(json.dumps(summary, indent=4))
        if 'Summary' in summary:
            precision = summary.get('Precision', 'Not Available')
            recall = summary.get('Recall', 'Not Available')
            print(f"Precision: {precision}")
            print(f"Recall: {recall}")

if __name__ == "__main__":
    main()


Evaluation Summary:
{
    "AggregatedEvaluationResults": {
        "ConfusionMatrix": [
            {
                "GroundTruthLabel": "Real",
                "PredictedLabel": "Real",
                "Value": 0.6666666666666666
            },
            {
                "GroundTruthLabel": "Real",
                "PredictedLabel": "Fake",
                "Value": 0.3333333333333333
            },
            {
                "GroundTruthLabel": "Fake",
                "PredictedLabel": "Real",
                "Value": 0.3333333333333333
            },
            {
                "GroundTruthLabel": "Fake",
                "PredictedLabel": "Fake",
                "Value": 0.6666666666666666
            }
        ],
        "F1Score": 0.8571428571428571,
        "Precision": 0.75,
        "Recall": 1.0
    },
    "EvaluationDetails": {
        "EvaluationEndTimestamp": "2024-08-01T23:11:25.388135",
        "Labels": [
            "Real",
            "Fake"
        ],
        "N

## 5 Teste Fim-a-Fim

Simule a operação fim-a-fim, com uma imagem de entrada forjada (foto de foto de um rosto) e outra com uma imagem de rosto, exibindo o resultado da classificação e a pontuação de cada classe.

In [23]:
import boto3

def detect_liveness(image_path, model_arn, min_confidence=95):
    client = boto3.client('rekognition')
    with open(image_path, 'rb') as image_file:
        image_bytes = image_file.read()

    response = client.detect_custom_labels(
        Image={'Bytes': image_bytes},
        ProjectVersionArn=model_arn,
        MinConfidence=min_confidence
    )

    return response['CustomLabels']

def main():
    model_arn = 'arn:aws:rekognition:us-east-1:011528301996:project/Faces/version/Faces.2024-08-01T19.59.45/1722553185343'

    fake_image_path = 'fake_1.jpg'
    real_image_path = 'real_1.jpeg'

    # Detecta vivacidade na imagem forjada
    fake_labels = detect_liveness(fake_image_path, model_arn)
    print("\nResultado para a imagem forjada:")
    if fake_labels:
        for label in fake_labels:
            print(f"Rótulo: {label['Name']}, Confiança: {label['Confidence']:.2f}")
    else:
        print("Nenhum rótulo detectado.")

    real_labels = detect_liveness(real_image_path, model_arn)
    print("\nResultado para a imagem real:")
    if real_labels:
        for label in real_labels:
            print(f"Rótulo: {label['Name']}, Confiança: {label['Confidence']:.2f}")
    else:
        print("Nenhum rótulo detectado.")

if __name__ == "__main__":
    main()



Resultado para a imagem forjada:
Rótulo: Fake, Confiança: 96.14

Resultado para a imagem real:
Rótulo: Real, Confiança: 95.53


>Com a implementação da solução na forma de uma aplicação do [Streamlit](https://www.streamlit.io/) (veja a pata streamlit-app e use o template) vale 1 ponto adicional.

**Pergunta**: Se utilizou o Streamlit, compartilhe a URL do aplicativo publicado:

**Resposta**:

## 6 Conclusões (2,5 pontos)

**Pergunta**: Dado todo o estudo e pesquisa, quais foram as conclusões sobre a solução, o que funcionou, o que não funcionou e quais os detalhes que observariam numa nova versão e melhorias do processo?

**Resposta**:
O experimento com Amazon Rekognition Custom Labels demonstrou eficácia em distinguir imagens reais de forjadas, com alta confiança: 96.14% para "Fake" e 95.53% para "Real". O modelo funcionou bem. Para melhorar, é interessante sempre aumentar a diversidade do dataset de treinamento. Além disso, a integração de detecção de anomalias e ajustes contínuos do modelo pode aprimorar sua robustez e adaptabilidade contra novas fraudes. Assim, o sistema pode se manter eficiente em cenários mais complexos.